src/gluonts/representation/representation.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+# Third-party imports
+from mxnet.gluon import nn
+import numpy as np
+
+# Standard library imports
+from typing import Tuple, Optional
+
+# First-party imports
+from gluonts.core.component import validated
+from gluonts.model.common import Tensor
+from gluonts.dataset.common import Dataset
+
+
+class Representation(nn.HybridBlock):
+    """
+        An abstract class representing input/output representations.
+
+        Parameters
+        ----------
+        is_output
+            Whether the representation is an input or output representation. It is necessary to make this distinction
+            because of different shaping considerations and optional child-block creation.
+            (default: False)
+    """
+
+    @validated()
+    def __init__(self, is_output: bool = False):
+        super().__init__()
+        self.is_output = is_output
+
+    def initialize_from_dataset(self, input_dataset: Dataset):
+        r"""
+        Perform some computation based on an entire dataset.
+
+        Parameters
+        ----------
+        input_dataset
+            GluonTS dataset.
+        """
+        pass
+
+    def initialize_from_array(self, input_array: np.ndarray):
+        r"""
+        Perform some computation based on a numpy array.
+
+        Parameters
+        ----------
+        input_array
+            Numpy array.
+        """
+        pass
+
+    # noinspection PyMethodOverriding
+    def hybrid_forward(
+        self,
+        F,
+        data: Tensor,
+        observed_indicator: Tensor,
+        scale: Optional[Tensor],
```

Comment:
In my experiments this did not prove to be an issue, not even with `hybridize=True`.

Reply:
Yes, I just checked: I think it's only an issue if you explicitly pass in or return a `None`. In this case if the input is not passed then the output is not `None` because the scale gets computed internally, so it's fine.

Reply:
It also seems to work when explicitly passing `None`. Calling hybrid_forward like this for example works fine in `SimpleFeedForwardEstimator` with both `hybridize=False` and `hybridize=True`:
```
input_tar_repr, scale = self.input_repr(
     past_target, F.ones_like(past_target), None
)
```

Reply:
Maybe it's a new thing: on MXNet 1.6 the following snippet

```python
import mxnet as mx

class MyNet(mx.gluon.HybridBlock):
    def hybrid_forward(self, F, x, y=None):
        return 2*x, y

net = MyNet()
net.initialize()
net.hybridize()

print(net(
    mx.nd.ones(shape=(10,)),
    None,
))
```

gives me

```
AssertionError: HybridBlock input must be (nested) list of Symbol or NDArray, but got None of type <class 'NoneType'>
```

Reply:
Correct me if I'm wrong, but I guess this is closer to a minimal working example for what we actually use and this works fine on my machine. Maybe you need to ensure that you actually pass a symbol object when having called `net.hybridize()` before?
```python
import mxnet as mx
from typing import Optional, Union, Tuple

Tensor = Union[mx.nd.NDArray, mx.sym.Symbol]

class MyNet(mx.gluon.HybridBlock):
    def hybrid_forward(self, F, x: Tensor, y: Optional[Tensor]
    ) -> Tuple[Tensor, Tensor]:
        return 2*x, y

net = MyNet()
net.initialize()

hybridize = True
if hybridize:
    F = mx.sym
    net.hybridize()
else:
    F = mx.nd

print(net(F.full(10,1),None))
```